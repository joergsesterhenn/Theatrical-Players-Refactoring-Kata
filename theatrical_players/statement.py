from theatrical_players.statement_formatter import StatementFormatter
from theatrical_players.plays_calculator import PlayCalculator


def statement(invoice, plays, statement_type="plain_text"):
    """
    Formats a statement for an invoice depending on  statement_type.
    :param invoice: invoice data
    :param plays: list of plays
    :param statement_type: format in which to render
           the statement (html, plain_text)
    :return: formated statement
    """
    entries = calculate_statement_entries(invoice['performances'], plays)
    total_amount = sum([entry.get("amount") for entry in entries])
    volume_credits = sum([entry.get("credits") for entry in entries])
    return StatementFormatter.of(statement_type).format_statement(
                            total_amount,
                            invoice["customer"],
                            entries,
                            volume_credits)


def calculate_statement_entries(performances, plays):
    """
    Calculates a list of entry dicts with all required information per play.
    :param performances: list of performances
    :param plays: list of plays
    :return: list of entry dicts with all required information per play
    """
    entries = []
    for performance in performances:
        play = plays[performance['playID']]
        play_calculator = PlayCalculator.of(play['type'])
        amount = play_calculator.calculate_amount(performance['audience'])/100
        volume_credits = (
            play_calculator.calculate_credits(performance['audience']))

        entries.append({'amount': amount, 'audience': performance['audience'],
                        'name': play["name"], 'credits': volume_credits})
    return entries
