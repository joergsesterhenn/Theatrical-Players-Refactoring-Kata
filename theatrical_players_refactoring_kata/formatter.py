def format_statement(entries: [], customer: str, amount: int,
                     volume_credits: int, statement_type="printout"):
    """
    :param entries: List of entry dicts
    :param customer: Name of the customer
    :param amount: Total amount owed
    :param volume_credits: Volume credit for invoice
    :param statement_type: how to render the statement (html/printout)
    :return: statement rendered according to type
    """
    if statement_type == "printout":
        return (f'Statement for {customer}\n'
                + format_printout_entries(entries)
                + f'Amount owed is {f"${amount:0,.2f}"}\n'
                + f'You earned {volume_credits} credits\n')
    elif statement_type == "html":
        return (f'<!DOCTYPE html>\n<html>\n<body>\n<h1>Statement for {
                customer}</h1>\n'
                + format_html_entries(entries)
                + f'<p>Amount owed is {f"${amount:0,.2f}"}</p>\n'
                + f'<p>You earned {volume_credits
                                   } credits</p>\n</body>\n</html>\n')
    else:
        raise Exception("Unknown type of statement: " + statement_type)


def format_printout_entries(entries):
    return "".join(
        [f' {entry.get("name")}: {f"${entry.get("amount"):0,.2f}"} ({
         entry.get("audience")} seats)\n' for entry in entries]
    )


def format_html_entries(entries):
    return "".join(
        [f'<p style="margin-left: 20px;">{
         entry.get("name")}: {f"${
          entry.get("amount"):0,.2f}"} ({
          entry.get("audience")} seats)</p>\n' for entry in entries]
    )
