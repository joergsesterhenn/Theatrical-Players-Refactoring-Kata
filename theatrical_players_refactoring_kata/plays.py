import math
from typing import Self
from abc import ABCMeta, abstractmethod


class Play(metaclass=ABCMeta):
    def __init__(self, play_type, play_name):
        self.play_type = play_type
        self.play_name = play_name

    @abstractmethod
    def calculate_amount(self, audience):
        pass

    @abstractmethod
    def calculate_credits(self, audience):
        pass

    @classmethod
    def of(cls, play_type, play_name) -> Self:
        if "comedy" == play_type:
            return ComedyPlay(play_type, play_name)
        elif "tragedy" == play_type:
            return TragedyPlay(play_type, play_name)
        else:
            raise ValueError(f'unknown type: {play_type}')


class ComedyPlay(Play):

    def calculate_amount(self, audience):
        this_amount = 30000
        if audience > 20:
            this_amount += 10000 + 500 * (audience - 20)
        this_amount += 300 * audience
        return this_amount

    def calculate_credits(self, audience):
        # add volume credits
        audience_volume_credits = max(audience - 30, 0)
        # add extra credit for every ten comedy attendees
        comedy_volume_credits = math.floor(audience / 5)
        return audience_volume_credits + comedy_volume_credits


class TragedyPlay(Play):
    def calculate_amount(self, audience):
        this_amount = 40000
        if audience > 30:
            this_amount += 1000 * (audience - 30)
        return this_amount

    def calculate_credits(self, audience):
        # add volume credits
        audience_volume_credits = max(audience - 30, 0)
        return audience_volume_credits
