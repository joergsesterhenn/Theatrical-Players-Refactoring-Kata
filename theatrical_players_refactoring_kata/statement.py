from theatrical_players_refactoring_kata.plays_calculator import PlayCalculator


def statement(invoice, plays, statement_type="printout"):
    entries = calculate_statement_entries(invoice, plays)
    total_amount = sum([entry.get("amount") for entry in entries])
    volume_credits = sum([entry.get("credits") for entry in entries])
    return format_statement(entries,
                            invoice,
                            total_amount,
                            volume_credits,
                            statement_type)


def calculate_statement_entries(invoice, plays):
    entries = []
    for perf in invoice['performances']:
        play_entry = plays[perf['playID']]
        play_calculator = PlayCalculator.of(play_entry['type'],
                                            play_entry['name'])
        amount = play_calculator.calculate_amount(perf['audience'])/100
        volume_credits = play_calculator.calculate_credits(perf['audience'])

        entries.append({'amount': amount, 'audience': perf['audience'],
                        'name': play_entry["name"], 'credits': volume_credits})
    return entries


def format_statement(entries, invoice, amount, volume_credits,
                     statement_type="printout"):
    if statement_type == "printout":
        return (f'Statement for {invoice["customer"]}\n'
                + format_printout_entries(entries)
                + f'Amount owed is {f"${amount :0,.2f}"}\n'
                + f'You earned {volume_credits} credits\n')
    else:
        return (f'<!DOCTYPE html>\n<html>\n<body>\n<h1>Statement for {
                invoice["customer"]}</h1>\n'
                + format_html_entries(entries)
                + f'<p>Amount owed is {f"${amount :0,.2f}"}</p>\n'
                + f'<p>You earned {volume_credits
                                   } credits</p>\n</body>\n</html>\n')


def format_printout_entries(entries):
    return "".join(
        [f' {entry.get("name")}: {f"${entry.get("amount") :0,.2f}"} ({
         entry.get("audience")} seats)\n' for entry in entries]
    )


def format_html_entries(entries):
    return "".join(
        [f'<p style="margin-left: 20px;">{entry.get("name")}: {f"${entry.get("amount") :0,.2f}"} ({
         entry.get("audience")} seats)</p>\n' for entry in entries]
    )



