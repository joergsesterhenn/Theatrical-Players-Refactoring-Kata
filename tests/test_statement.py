import json

import pytest
from approval_utilities.utils import get_adjacent_file
from approvaltests import verify
from theatrical_players_refactoring_kata.statement import statement


def test_example_statement():
    with open(get_adjacent_file("invoice.json")) as f:
        invoice = json.loads(f.read())
    with open(get_adjacent_file("plays.json")) as f:
        plays = json.loads(f.read())
    result = statement(invoice, plays)
    verify(result)


def test_statement_with_new_play_types():
    with open(get_adjacent_file("invoice_new_plays.json")) as f:
        invoice = json.loads(f.read())
    with open(get_adjacent_file("new_plays.json")) as f:
        plays = json.loads(f.read())
    result = statement(invoice, plays)
    verify(result)


def test_statement_with_new_play_types_and_html():
    with open(get_adjacent_file("invoice_new_plays.json")) as f:
        invoice = json.loads(f.read())
    with open(get_adjacent_file("new_plays.json")) as f:
        plays = json.loads(f.read())
    result = statement(invoice, plays, "html")
    verify(result)
