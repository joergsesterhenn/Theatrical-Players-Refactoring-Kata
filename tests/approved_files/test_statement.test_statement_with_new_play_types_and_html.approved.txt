<!DOCTYPE html>
<html>
<body>
<h1>Statement for BigCoII</h1>
<p style="margin-left: 20px;">Henry V: $630.00 (53 seats)</p>
<p style="margin-left: 20px;">As You Like It: $650.00 (55 seats)</p>
<p>Amount owed is $1,280.00</p>
<p>You earned 48 credits</p>
</body>
</html>
